concurrency/PubSubMessageService.java

```diff
@@ -0,0 +1,389 @@
+                }
+            }
+
+            // remove from the listener -> topic handlers map
+            Collection<TopicEventHandler> topicHandlers = topicHandlersMap.get(clazz);
+            if (topicHandlers != null) {
+                // now remove all event handlers declared in this listener from the
+                // topic -> handlers map.
+                for (TopicEventHandler topicHandler : topicHandlers) {
+                    topicHandler.unregisterListener(listener);
+                }
+            }
+        }
+    }
+
+    /**
+     * remove the handler if the target object (the subscriber) has been
+     * de-referenced (gc-ed)
+     *
+     * @param handler
+     */
+    void removeHandler(BaseEventHandler handler) {
+        if (!handler.hasListener()) {
+            if (handler instanceof EventHandler) {
+                EventHandler typeEventHandler = (EventHandler) handler;
+                Class<?> eventType = typeEventHandler.getEventType();
+                Collection<EventHandler> handlers = eventTypeToHandlers.get(eventType);
+                if (handlers != null) {
+                    handlers.remove(handler);
+                    if (handlers.isEmpty()) {
+                        eventTypeToHandlers.remove(eventType);
+                    }
+                }
+            } else if (handler instanceof TopicEventHandler) {
+                TopicEventHandler topicEventHandler = (TopicEventHandler) handler;
+                String topic = topicEventHandler.topic;
+                Collection<TopicEventHandler> handlers = topicToHandlers.get(topic);
+                if (handlers != null) {
+                    handlers.remove(handler);
+                    if (handlers.isEmpty()) {
+                        topicToHandlers.remove(topic);
+                    }
+                }
+            }
+        }
+    }
+
+    ExecutorService getMessageQueueExecutor() {
+        if (centralExecutor) {
+            return eventExecutor;
+        } else {
+            return newExecutor();
+        }
+    }
+
+    @Nonnull
+    private static ExecutorService newExecutor() {
+        return new ThreadPoolExecutor(DEFAULT_MIN_POOL_SIZE,
+                DEFAULT_MAX_POOL_SIZE, DEFAULT_THREAD_IDLE_TIMEOUT,
+                TimeUnit.SECONDS, new LinkedBlockingDeque<Runnable>(),
+                new NamedThreadFactory("pubsub-" + executorNumber.getAndIncrement()));
+    }
+
+    void shutdownExecutor(ExecutorService executor) {
+        if (executor == null || executor == eventExecutor) {
+            return;
+        } else {
+            executor.shutdown();
+            try {
+                // wait for 100ms and then force shutdown the executor
+                if (!executor.awaitTermination(
+                        DEFAULT_THREAD_POOL_SHUTDOWN_TIMEOUT,
+                        TimeUnit.MILLISECONDS)) {
+                    executor.shutdownNow();
+                }
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)